mod.ts

```diff
@@ -0,0 +1,81 @@
+const canElideFrames = 'captureStackTrace' in Error
+const startStackFrames = new WeakMap()
+
+interface Result {
+  name: 'AssertionError' | 'AssertionResult'
+  ok: boolean
+  toJSON(...args: unknown[]): Record<string, unknown>
+}
+
+export class AssertionError<T> extends Error implements Result {
+  [key: string]: unknown
+
+  get name(): 'AssertionError' {
+    return 'AssertionError'
+  }
+
+  get ok() {
+    return false
+  }
+
+  constructor(public message = 'Unspecified AssertionError', props?: T, ssf?: Function) {
+    super(message)
+    if (canElideFrames && ssf) startStackFrames.set(this, ssf)
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
+  }
+
+  get stack() {
+    if (canElideFrames) {
+      return (Error as any).captureStackTrace(this, startStackFrames.get(this) || AssertionError);
+    } else {
+      return super.stack
+    }
+  }
+
+  toJSON(stack: boolean): Record<string, unknown> {
+    return {
+      ...this,
+      message: this.message,
+      ok: false,
+      // include stack if exists and not turned off
+      stack: stack !== false ? this.stack : undefined,
+    }
+  }
+
+}
+
+export class AssertionResult<T> implements Result {
+  [key: string]: unknown
+
+  get name(): 'AssertionResult' {
+    return 'AssertionResult'
+  }
+
+  get ok() {
+    return true
+  }
+
+  constructor(props?: T, ssf?: Function) {
+    for (const key in props)
+      if (!(key in this))
+        // @ts-ignore
+        this[key] = props[key];
+  }
+
+  toJSON(): Record<string, unknown> {
+    return {
+      ...this,
+      ok: this.ok,
+    }
+  }
+}
+
+const x = new AssertionError<{has: true}>('hi', {has:true})
+x.has === true
+
+export const ok = <T extends {}>(val: T) => new AssertionResult<T>(val)
+export const error = <E extends {}>(message: string, val: E) => new AssertionError<E>(message, val)
+export default { ok, error }
```

Comment:
This is interesting. What's the benefit of doing this as opposed to importing `AssertionError` and `AssertionResult`?

Reply:
It’s just a shorthand. Fine to remove it but thought it might be nice to have.